flink-connectors/flink-connector-kafka/src/main/java/org/apache/flink/streaming/connectors/kafka/table/KafkaDynamicSource.java

```diff
@@ -329,11 +375,101 @@ public int hashCode() {
                 specificStartupOffsets,
                 startupTimestampMillis,
                 upsertMode,
+                tableIdentifier,
+                fallbackToFlinkKafkaConsumer,
                 watermarkStrategy);
     }
 
     // --------------------------------------------------------------------------------------------
 
+    protected KafkaSource<RowData> createKafkaSource(
+            DeserializationSchema<RowData> keyDeserialization,
+            DeserializationSchema<RowData> valueDeserialization,
+            TypeInformation<RowData> producedTypeInfo) {
+        final MetadataConverter[] metadataConverters =
+                metadataKeys.stream()
+                        .map(
+                                k ->
+                                        Stream.of(ReadableMetadata.values())
+                                                .filter(rm -> rm.key.equals(k))
+                                                .findFirst()
+                                                .orElseThrow(IllegalStateException::new))
+                        .map(m -> m.converter)
+                        .toArray(MetadataConverter[]::new);
+
+        // check if connector metadata is used at all
+        final boolean hasMetadata = metadataKeys.size() > 0;
+
+        // adjust physical arity with value format's metadata
+        final int adjustedPhysicalArity =
+                producedDataType.getChildren().size() - metadataKeys.size();
+
+        // adjust value format projection to include value format's metadata columns at the end
+        final int[] adjustedValueProjection =
+                IntStream.concat(
+                                IntStream.of(valueProjection),
+                                IntStream.range(
+                                        keyProjection.length + valueProjection.length,
+                                        adjustedPhysicalArity))
+                        .toArray();
+
+        final KafkaDeserializationSchema<RowData> kafkaDeserializer =
+                new DynamicKafkaDeserializationSchema(
+                        adjustedPhysicalArity,
+                        keyDeserialization,
+                        keyProjection,
+                        valueDeserialization,
+                        adjustedValueProjection,
+                        hasMetadata,
+                        metadataConverters,
+                        producedTypeInfo,
+                        upsertMode);
+
+        final KafkaSourceBuilder<RowData> kafkaSourceBuilder = KafkaSource.builder();
+
+        if (topics != null) {
+            kafkaSourceBuilder.setTopics(topics);
+        } else {
+            kafkaSourceBuilder.setTopicPattern(topicPattern);
+        }
+
+        // For compatibility with legacy source that is not validating group id
+        if (!properties.containsKey(ConsumerConfig.GROUP_ID_CONFIG)) {
+            kafkaSourceBuilder.setGroupId("KafkaSource-" + tableIdentifier.asSummaryString());
```

Comment:
shall we add a warning to the log to inform the user that the DDL needs to be updated?